### DISCUSSION TITLE
Documenting exposed commands

### DISCUSSION THREAD

**Discussion by elParaguayo** [qtile/qtile#3016] [General]
@m-col This isn't ready for a PR (not even a draft one) but it would be good if you could take a look at this branch (https://github.com/elParaguayo/qtile/tree/docs-lazy) and let me know your thoughts.

Issues that I'm aware of:
- [ ] Some missing commands (this is because I try to filter out inherited commands to prevent every widget showing the CommandObject commands)
- [ ] Needs better headings
- [ ] Formatting of docstring in the table is **ugly**

**Comment by m-col** [comment]
Are there specific problems with the current state that you're trying to tackle?

> (this is because I try to filter out inherited commands to prevent every widget showing the CommandObject commands)

This is definitely annoying for the widgets, but do we want filtering for everything else? E.g. `info` or `function` can be used anywhere and could be useful for people to know.

**Reply by elParaguayo** [reply]
They're not hidden. They're displayed first and then only those widgets that add new commands are shown. 

You may want to build docs locally to see what I mean.

**Reply by elParaguayo** [reply]
What I wanted to do was show `CommandObject` commands at the beginning of each object's section, but examples tailored accordingly e.g. `lazy.window.eval`, `lazy.widget["widgetname"].eval` etc.

If I don't split out other inherited commands (e.g. `libqtile.base.widget._Widget.cmd_info`) then every widget will be listed - but maybe that's ok here.

**Reply by m-col** [reply]
(Warning, this became a bit of a stream-of-consciousness on many aspects of the docs :laughing: )

In general I like the change, we do need a comprehensive API reference of the commands. A couple thoughts (which you're likely aware of / already plan to do):

- The `CommandObject` tables are more intrusive than I thought, when looking at the build. Perhaps we can render a single copy of the table at the top and clarify that they apply to everything. We could also avoid writing examples for each one too.
- The X11 Core doesn't have any unique commands, and the Wayland core has its commands listed on the Wayland-specific page; perhaps we can exclude the cores from that tables?
- We have a Built-in Layouts section that describes their configuration. I wonder if we can add their commands there, keeping that stuff together?
- Same as above but with widgets
- "Commands API" and "Scripting Commands" are two sides of the same coin. What would be nice is if the content from "Commands API" formed a kind of introduction to the commands, and then following that we list what is available.
- Similarly the "Scripting" tab could probably come on what is now the "Scripting Commands" page.
- In many places (not just in the files touched here but in the docs in general) the contents sidebar seems to show too much IMO and it makes it seem very clutted and difficult to digest. E.g. the "Commands API" page could be a single page with a single entry in that contents page, which would also make it fit more nicely as an intro page to the API reference. I assume this is large part default behaviour but I hope we can control that a bit. As an extreme example of what sphinx likes to do, this is from the commands list, and it could just as well end at "Backend" here I think:
![image](https://user-images.githubusercontent.com/44536113/141661687-eee95e58-5b6d-4599-ba93-b801de29a4ae.png)

- The examples at config/lazy.html seem redundant with the extensive reference that they link too. Do you think its necessary to have examples there?
- The command reference and "Commands API" info page might be better coming before the "Built-In X" pages. They're certainly more important than the extensions which come first. If the layouts and widgets came after the commands reference and they had their commands listed on their "Built-In Layouts/Widgets" page then we've already explained the lazy commands stuff which makes more sense flow-wise.
- I'm not sure I like the tables introduced at the start of each section. We have the `Keys` section on the "Commands API" page which I think nicely describes how some lazy paths optionally take keys e.g. `group['"name"]`. Similarly we should ensure that the different ways these can be accessed (shell, qshell, `lazy` etc) are made clear so we don't have to list them each time.
- A common pattern in many of these thoughts and my general criticism of our docs in their current form is how they are incredibly verbose and can be difficult to navigate. A lot of effort has gone into writing docs which is _fantastic_, but we could probably make things easier for users by shuffling some of it around, removing redundancies, and merging things that overlap in scope. It would naturally follow that our contents sidebar would shrink a bit and it would be easier for us to say "check out the page on X in the docs" because as it stands it's difficult to do (for me at least!). I get lost in our docs, and I'm very familiar with the code, so that cannot be a good sign :laughing:

**Reply by elParaguayo** [reply]
Thank you so much for taking the time to look at this properly. This is exactly why I didn't submit as a draft PR!

Lots of great points, many of them are things that bugged me too. I'll respond a bit more fully later this morning.

**Reply by elParaguayo** [reply]
> * The `CommandObject` tables are more intrusive than I thought, when looking at the build. Perhaps we can render a single copy of the table at the top and clarify that they apply to everything. We could also avoid writing examples for each one too.

Yes - Having seen how it looks on the page, I agree. Might be worth adding some explanations too e.g. how `lazy.object.function` passes the object as the first argument.

> * The X11 Core doesn't have any unique commands, and the Wayland core has its commands listed on the Wayland-specific page; perhaps we can exclude the cores from that tables?

It has `info`. Admittedly, no one is likely to use it, but it is exposed so I would like to list it. Now, if _everything_ has an `info` command then we could add to `CommandObject` and then just filter it out from everything else.

> * We have a Built-in Layouts section that describes their configuration. I wonder if we can add their commands there, keeping that stuff together?

Yes. This is also what I was thinking. Having config and command details on the same page make a lot of sense but I also like the idea of a single reference page of all commands. No reason not to do both.

> * Same as above but with widgets

Same as above but with widgets :wink:

> * "Commands API" and "Scripting Commands" are two sides of the same coin. What would be nice is if the content from "Commands API" formed a kind of introduction to the commands, and then following that we list what is available.

Yes. It feels like there's a broader rewrite needed here.

> * Similarly the "Scripting" tab could probably come on what is now the "Scripting Commands" page.

:+1: 

> * In many places (not just in the files touched here but in the docs in general) the contents sidebar seems to show too much IMO and it makes it seem very clutted and difficult to digest. E.g. the "Commands API" page could be a single page with a single entry in that contents page, which would also make it fit more nicely as an intro page to the API reference. I assume this is large part default behaviour but I hope we can control that a bit. As an extreme example of what sphinx likes to do, this is from the commands list, and it could just as well end at "Backend" here I think:
>   ![image](https://user-images.githubusercontent.com/44536113/141661687-eee95e58-5b6d-4599-ba93-b801de29a4ae.png)

Yes, I didn't love this either but was a result of having to split X11 and Wayland.

> * The examples at config/lazy.html seem redundant with the extensive reference that they link too. Do you think its necessary to have examples there?

Possibly not. Need to think about that one.

> * The command reference and "Commands API" info page might be better coming before the "Built-In X" pages. They're certainly more important than the extensions which come first. If the layouts and widgets came after the commands reference and they had their commands listed on their "Built-In Layouts/Widgets" page then we've already explained the lazy commands stuff which makes more sense flow-wise.

That could work.

> * I'm not sure I like the tables introduced at the start of each section. We have the `Keys` section on the "Commands API" page which I think nicely describes how some lazy paths optionally take keys e.g. `group['"name"]`. Similarly we should ensure that the different ways these can be accessed (shell, qshell, `lazy` etc) are made clear so we don't have to list them each time.

:+1: They came out messier than I would have liked. I did want to give actual examples of how the objects could be accessed via each method to be as helpful as possible. Will rethink.

> * A common pattern in many of these thoughts and my general criticism of our docs in their current form is how they are incredibly verbose and can be difficult to navigate. A lot of effort has gone into writing docs which is _fantastic_, but we could probably make things easier for users by shuffling some of it around, removing redundancies, and merging things that overlap in scope. It would naturally follow that our contents sidebar would shrink a bit and it would be easier for us to say "check out the page on X in the docs" because as it stands it's difficult to do (for me at least!). I get lost in our docs, and I'm very familiar with the code, so that cannot be a good sign laughing

:+1:  Yes, the docs looks as though they've been built from multiple different commits rather than being a cohesive site.

**Reply by elParaguayo** [reply]
> We have a Built-in Layouts section that describes their configuration. I wonder if we can add their commands there, keeping that stuff together?

Worth adding that the current docs use the `:no-commands:` directive to hide the commands on those pages. If you remove that directive you can see what it looks like.

I don't love it but it's easier enough to replace that format with the template I'm working on (which excludes `CommandObject` commands).

![20211114_184234](https://user-images.githubusercontent.com/946265/141694220-2a14cec5-db59-4f95-8402-f1d7bb1368b4.png)

We can ignore the typo for now!!

**Reply by elParaguayo** [reply]
The current formatting is like this:

![20211114_184740](https://user-images.githubusercontent.com/946265/141694354-f3d193ba-5517-49c8-8d39-780d29b2256f.png)

Maybe we should stick with that. However, I would like to show it formatted as a lazy command too.

Thoughts?

**Reply by m-col** [reply]
I agree that the formatting could be improved and I like how you were doing it. Most users will only know the lazy interface and having that everywhere (e.g. now it being usable in widget mouse_buttons) it's nice to have that as the "official API". I don't think we even need to document the `cmd_` stuff - and even better when they don't have that prefix, as we can just tell people they can access the lazy objects at the same named method.

**Reply by elParaguayo** [reply]
Agreed. But the advantage of the other way is the formatting of the docstring.

**Reply by m-col** [reply]
Because it's rendered markup? We could likely still parse it similarly before it enters the table.

**Reply by elParaguayo** [reply]
That's the bit I'm struggling with. Came up with errors as there were section titles in there which can't go inside the table (as far as I can tell). Docstring is therefore escaped first before going into the table.

**Comment by m-col** [comment]
Another thing I just noticed: on the `Keys` page at http://docs.qtile.org/en/latest/manual/config/keys.html it explains `lazy.X.when()` but this actually is general and applies to all lazy objects so should be moved elsewhere.

**Reply by elParaguayo** [reply]
I think I added that. Maybe I put it in the wrong place!

**Reply by m-col** [reply]
I have the memory of a goldfish bowl :)
